### PR TITLE
[fuzz] Use public methods in addrman fuzz tests

### PR DESCRIPTION
**Pull request by jnewbery**
#22974 improved the performance of `CAddrMan::Good()` significantly so that it could be used directly in the fuzz tests, instead of those tests reaching directly into addrman's internal data structures.

This PR continues the work of making the fuzz tests only use `CAddrMan`'s public methods and pulls the `Fill()` and `RandAddr()` methods from `CAddrManDeterministic` into free functions, since they no longer need access to `CAddrMan` internals.

### CODE REVIEW

**Comment by jnewbery** [comment]
@vasild - this addresses your review comment here: https://github.com/bitcoin/bitcoin/pull/22974#pullrequestreview-757542948

File: src/test/fuzz/addrman.cpp [hunk 1]
```diff
@@ -23,92 +24,85 @@ void initialize_addrman()
     SelectParams(CBaseChainParams::REGTEST);
 }
 
-class CAddrManDeterministic : public CAddrMan
+/** Generate a random address. Always returns a valid address. */
+static CNetAddr RandAddr(FuzzedDataProvider& fuzzed_data_provider, FastRandomContext& fast_random_context)
+    EXCLUSIVE_LOCKS_REQUIRED(cs)
```
Comment (MarcoFalke) [inline src/test/fuzz/addrman.cpp:29]: leftover annotation?

Comment (jnewbery) [reply src/test/fuzz/addrman.cpp]: Oops! Fixed now. Thank you.

**Comment by jnewbery** [comment]
Rebased

File: src/test/fuzz/addrman.cpp [hunk 2]
```diff
@@ -34,92 +35,84 @@ FUZZ_TARGET_INIT(data_stream_addr_man, initialize_addrman)
...
 
-            for (size_t j = 0; j < num_addresses; ++j) {
-                const auto addr = CAddress{CService{RandAddr(), 8333}, NODE_NETWORK};
-                const auto time_penalty = insecure_rand.randrange(100000001);
-                Add_(addr, source, time_penalty);
+    return addr;
+}
 
-                if (n > 0 && mapInfo.size() % n == 0) {
-                    Good_(addr, false, GetTime());
```
Comment (vasild) [inline src/test/fuzz/addrman.cpp:113]: The new code calls `Good()` which calls `Good_(/*test_before_evict=*/true)` whereas the old code passed `false`. I think this is ok.

**Review by vasild — APPROVED** [review]
ACK 19e2d94ecd599e469e3ff168b456088e12e61f6f

The public `Good()` will do some extra locking and call `Check()` in comparison to `Good_()`. The locking should be insignificant wrt performance and the `Check()` should be noop due to the used `consistency_check_ratio=0`. Also, the public `Good()` will do "test before evict" which could cause some slowdown.

I did measure the timings to fill an addrman with hardcoded `n=3 (33% tried)`, `num_sources=50` and `num_addresses=500` on both `master` and this PR. The addrman ends up with about 16k addresses.

`master`:
```
   13.493194 sec: fill
    0.534130 sec: ser
    5.664423 sec: deser
    1.330460 sec: compare

   13.488360 sec: fill
    0.529679 sec: ser
    5.632100 sec: deser
    1.321111 sec: compare
```

This PR:

```
   14.130295 sec: fill
    0.553457 sec: ser
    5.809201 sec: deser
    1.392843 sec: compare

   13.757125 sec: fill
    0.556877 sec: ser
    5.863635 sec: deser
    1.371259 sec: compare
```

**Comment by practicalswift** [comment]
Concept ACK

**Comment by jnewbery** [comment]
Rebased on #22950. Leaving as a draft until that PR is merged.

**Comment by MarcoFalke** [comment]
Rebase and fix CI errors?

**Comment by jnewbery** [comment]
Rebased. Marking ready for review.

**Comment by MarcoFalke** [comment]
```
SUMMARY: ThreadSanitizer: data race /usr/lib/llvm-12/bin/../include/c++/v1/ios:523:12 in std::__1::ios_base::width() const

**Comment by amitiuttarwar** [comment]
concept ACK

**Comment by theuni** [comment]
utACK 44452110f0fa7cc1bcb941a3c7b5db4b492a7b9c. Agree the failure seems unrelated, looks like some startup race.

Thanks very much @vasild for the benchmarks. Indeed it appears the only functional change here is hitting more locks and `test_before_evict` is true, good to know those don't slow things down too much.

It appears to me that AddrManDeterministic would no longer be required if `AddrMan` accepted a reference to a `FastRandomContext` and grew an equality operator, though I'm not sure there's any need for that.

**Review by vasild — APPROVED** [review]
ACK 44452110f0fa7cc1bcb941a3c7b5db4b492a7b9c

**Pull request merged**